### ISSUE TITLE
🚀 Feature: Configurable rate limits

### ISSUE THREAD

**Issue by ameerfayiz** [appwrite/appwrite#2953]
### 🔖 Feature description

instead of global Rate Limits of 60 requests in every 1 minutes per IP address. , there should be a provision to manually change that 60 request per minute limit.

### 🎤 Pitch

 Because some functions should be allowed to execute 2 or 3 execution per minute like for OTP resending, etc..  now I have to manually control the limit by creating a proxy function and its hard  to manage for functions to have different rate limits, even though possible to do it manually.

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by eldadfux** [comment]
This has been raised internally in the context of collections and document creation. We'd love to gather more community feedback before we push this feature to work.

**Comment by ameerfayiz** [comment]
im using tokenbucket algorithm inside all cloud functions which need individual rate limits to achieve the individual rate limit effect. which needs an extra table with 2 attributes for every such cloud functions.

**Comment by DankMemeGuy** [comment]
I agree on configurable rate limit. Servers are not all the same, and can handle different loads. Different websites have different levels of protection they really need, as well. Making this option configurable via UI would be a good solution. 

If protection is really at front of mind, you could enable an 'alert' feature that triggers when a certain threshold is reached, whether site wide or per user/IP. Such as a user hitting 80% of their rate limit or something. That way an admin can just deal with the abuse when it happens, or see what abuse just occurred when they check the alerts next, or have the ability to easily impose a tighter limit if the website is undergoing an event.

**Comment by d4rckh** [comment]
This would a must-have for apps that have large amounts of functions

**Comment by huy-lv** [comment]
Hope this feature will be launch soon

**Comment by delqu** [comment]
I also waiting for this feature

**Comment by JoshiJoshiJoshi** [comment]
Is there an ETA when this is going to be worked on?

**Comment by webpatser** [comment]
I had a demo where 15 people tried out a new app. All on the same wifi. So rate limit was triggered for signup because they shared the IP address. This is rate limit should really be adjustable.

```
This endpoint is limited to 10 requests in every 60 minutes per IP address. We use rate limits to avoid service abuse by users and as a security practice. [Learn more about rate limiting](https://appwrite.io/docs/rate-limits).
```

**Comment by khare4** [comment]
> This has been raised internally in the context of collections and document creation. We'd love to gather more community feedback before we push this feature to work.

this is must have feature so that people can set rate limit according to there need.

**Comment by jhooper04** [comment]
I think rate limiting would vary wildly on a per project basis of what is normal use of the system and what is abuse. `_APP_OPTIONS_ABUSE=disabled` with your own rate limiting proxy is a good approach but seems like that is going to be nearly every project that needs to do that beyond simple configuration type projects.

One of the great things I like about appwrite is that it is an all in one package to get an app up quickly and still be able to scale it later. Lots of features out of the box. Makes sense to me that if it is going to offer a rate limiting feature, it should be configurable per route; otherwise, most devs will have to disable it and implement that elsewhere. avoiding as much extra work as possible is what drew me to appwrite personally.

**Comment by sgaabdu4** [comment]
I'm not able to tweak the rate limiting for the cloud as we don't have access to .env or is there another way?

**Comment by joeyouss** [comment]
Hi @sgaabdu4 yes you cannot edit env on cloud yet but this is something we can probably discuss, however which rate limit you want to change? We might have some solution with Appwrite functions

**Comment by DHsoftw** [comment]
> > This has been raised internally in the context of collections and document creation. We'd love to gather more community feedback before we push this feature to work.
> 
> this is must have feature so that people can set rate limit according to there need.

True. Opened an issue. Give it a 👍

**Comment by DHsoftw** [comment]
Also, It would be great to have the possibility to modify other abuse limits like pagination or max document fetched limits.
That could prevent anyone from fetching 5000 documents in an app where it's unnecessary fetching more than 30 documents with one query.

**Comment by jhooper04** [comment]
I think the functions would be the main need for custom rate limits since the function could do just about anything. Though you could implement that in the function itself. It would definitely be nice to be able to set a rate on any endpoint and simplify the functions.

**Comment by DHsoftw** [comment]
Yes, the problem is that fetching documents should not be done through functions, since that's translated into a slower app and will need a more powerful server (and more expensive)

**Comment by FranciscoKloganB** [comment]
This might not be the usual Appwrite flow, but my app's registration process crucial steps other than creating a simple account itself. Consequently, registration and also login acts, are made on the server side, within Nuxt. Not directly from the WebSDK to appwrite. With 10 requests per 60m for createAccount and 10 requests per 60m for createEmailSession, I can not have more than 10 registrations per hour or more than 10 logins. :mechanic: 

I am not willing to offload this funcionality to the client/browser. If appwrite is not `Fullstack Framework` friendly, I am afraid I will need to find other solutions to put my money on. :\

**Comment by DHsoftw** [comment]
@FranciscoKloganB For server side applications you need to use a server side SDK and an API key so you will not see any rate limit

**Comment by FranciscoKloganB** [comment]
@DHsoftw that is true, but the `Appwrite Node SDK` does not expose `createAccount`, it exposes at best `createEmailSession` if you are able to generate a JWT on the client side, but then, I am unable to set the cookie domain as intended on the server side.

**Comment by TendieShop** [comment]
How do I implement _APP_OPTIONS_ABUSE = disabled in my nextjs app, I don't see any documentation on that. I have tried putting it in .env and also .env.local, but nothing seems to work, please help...

**Comment by TendieShop** [comment]
How do I implement _APP_OPTIONS_ABUSE = disabled in my nextjs app, I don't see any documentation on that. I have tried putting it in .env and also .env.local, but nothing seems to work, please help @stnguyen90

**Comment by neoighodaro** [comment]
> How do I implement _APP_OPTIONS_ABUSE = disabled in my nextjs app, I don't see any documentation on that. I have tried putting it in .env and also .env.local, but nothing seems to work, please help @stnguyen90

It's for self hosted instances of appwrite.

**Comment by TendieShop** [comment]
> > How do I implement _APP_OPTIONS_ABUSE = disabled in my nextjs app, I don't see any documentation on that. I have tried putting it in .env and also .env.local, but nothing seems to work, please help @stnguyen90
> 
> It's for self hosted instances of appwrite.

I see, thank you

**Comment by dhruv-tech** [comment]
Expressing my interest for this feature as well. The default limits are just WAY too low, especially for apps that are going to be used by multiple concurrent users over the same wifi network.

**Comment by jspm2013** [comment]
dont know what an effort to implement and test it would be, but wouldnt it be most effective if youd implement a solution where each dev could configure a .env-variable on a per restricted-topic base (self-hosted), or probably set it somewhere in the cloud-dashboard (cloud)?!

I mean for sure this would be only suitable if the amount of topics which are restricted by the _APP_OPTIONS_ABUSE=enabled setting are just a few, lets say below 10 - but as far as I understood, it is something like this...meaning they are not like 30 different topics which get controlled by this switch?!

At the end, if youd provide such a detailed control over this rate limits, it would be a very strong security/DX feature and selling point I guess, and probably - I dont know - it would be not that much complexity to implement this...?

**Comment by DerLeole** [comment]
Hey, so I'd like to pick this topic up again for discussion, as its impacting my current project.

I spent a good bit of time to look into the mechanisms of rate limiting within appwrite and came up with some ideas that I would like to float with the maintainers and anyone who might make use of such a feature.

## How appwrite rate limiting currently works
The code that evaluates rate limiting is located almost entirely within [app/controllers/shared/api.php](https://github.com/appwrite/appwrite/blob/main/app/controllers/shared/api.php) (currently line 329+).

There are 3 main criteria to every rate limit evaluation:
- _abuse-limit_
  - How many times you are allowed to access something in a given time frame.
- _abuse-time_
  - The time window length, after which the rate limit resets.
  - The algorithm used is Fixed Window
- _abuse-key_
  - How appwrite differentiates between individual api accesses and logs their usage for rate limit purposes
  - I will ignore this for now, as I think this goes a bit beyond the scope of what is needed and could be far too easily misconfigured.
  - The hardcoded defaults here are also good.

All of these main criteria are either set on a per route basis in the route definition using `->label()`, or if no label is set, a default value (abuse-limit 0, so off, and abuse-time 3600). Funnily enough some routes already use a default set by a hardcoded variable in init.php that determines the max write rate.

If any rate limit is found to be exceeded, an exception is thrown/returned.
The final check is overridden in any of the following cases:
- The requesting user is an API key
- The requesting user is priviliged, meaning he is admin, developer, or owner in the project in the appwrite console (only works for console users)
- The `_APP_OPTIONS_ABUSE` environment variable is set to `disabled`

## What should be configurable?
Before anything can be implemented, I want to find a solution that works for most people that struggle with rate limits and that fits the architecture of appwrite. I have some ideas, but would like some feedback.

### [1] Global override
Users can configure a global override for limit and time, which overrides any defaults specified inside the routes. Pretty inflexible and not very configurable but super easy to implement.

### [2] Global modifier
User can configure a modifier like `+100` or `x2` for limit and time, which is applied to any defaults set inside the routes. A bit more flexible, but still modifies all routes.

### [3] Per route & method override
Pretty much the same as the global override, but you can specify which route and method it would apply to. The routes from the REST documentation could be used as the key.
This would be pretty much the best possible version of this feature, but is of course much more complicated to implement than any global solutions.

## How to implement?
Should adjustable rate limiting be something that is a full feature, configurable in the appwrite console, or is it better to leave it as something configured by IT in the environment variables?

### Environment variables
The big pro here is the easy and swiftness of implementation. The feature could be implemented entirely in  [app/controllers/shared/api.php](https://github.com/appwrite/appwrite/blob/main/app/controllers/shared/api.php) with a couple of lines of code, depending on the granularity. It would also fit in line with the existing env option to disable rate limiting.

The downsides are clear: It would be unavailable for appwrite cloud users and any changes apply to all projects of the instance.
Env vars are also not really suited to change a whole bunch of granular options at once. You'd have to either have a small number of pre-defined envs with a potentilly large and hard to read number of options in them (paths are long), or you'd have to make use of parametric env names (i.e `_APP_ABUSE_LIMIT_GET_databasesdatabaseIdcollectionscollectionIddocuments`).

### Console implementation
Adjust rate limits in the project settings using a granular configuration menu, similar to permissions. Set an override for specific routes and methods on a per-project basis. Potentially even only for specific collections, databases, users, teams etc (though that would probably be too much for now).

The downside is of course the scale of this feature. It would require a new collection in the projectDb, new API routes, new UI elements, and documentation, testing etc. Compared with the couple lines of code in one file the env based solution requires, thats a big undertaking for a feature, that not many might even use.


Sorry for the wall of text, I really wanted to reignite this discussion again and offer some concrete ideas and a starting point for either me or someone else to go on.

**Comment by stnguyen90** [comment]
Related discussion: https://github.com/appwrite/appwrite/discussions/7509

**Comment by xyvielyons** [comment]
i myself i would like a configurable rate limit especially in development when setting up authentication

**Comment by eldadfux** [comment]
We look to solve this problem in two main ways. First solution is development keys, these keys will be time limited to avoid accidental usage of them in a production environment. We aim to release this in Appwrite 1.7. As @stnguyen90 pointed out more information about this solution can be found here: https://github.com/appwrite/appwrite/discussions/7509

This solution is designed to solve the problem with rate limits that are designed against brute force attack, especially on auth related endpoints.

2nd solution which is a bit more in early stages of discussion is the Appwrite Firewall. This is basically the idea of having something that is more similar to an L7 firewall aka WAF. This will allow greater level of flexibility in setting custom rules across all of your Appwrite endpoints.

**Comment by daveycodez** [comment]
Ahh I almost left Supabase for your built in rate limiting, but 60 per IP is too low sadly.

**Comment by Amitrochates** [comment]
much needed

**Comment by LorenzoBloedow** [comment]
I ran into this issue and created an open-source library to handle this, it supports 4 different algorithms, multiple strategies at the same time, allows you to self-host or use the rate limiting function inline for less latency, and a bunch of other options, fully typed with JSDoc.

I have plans in the future to have a paid managed option too, but I don't have one at the time of writing, so I wonder if it's ok to link it here so people can use the free version, I really don't want to spam or hijack this issue.

Thanks :)

**Comment by shlokstoira** [comment]
@LorenzoBloedow please share

**Comment by LorenzoBloedow** [comment]
@shlokstoira
 
Here you go:
https://github.com/borrowdev/borrow

Hopefully it helps. :)

If someone from Appwrite doesn't want me to share this please let me know and I'll delete all comments.
Again, don't want to hijack this issue.

**Comment by hortigado** [comment]
I need it for suppot offline, the upload of data is very slow when the user connect to the red by it limit

**Comment by ammarfaris** [comment]
Is there any update on this? just wondering

**Comment by ammarfaris** [comment]
Why is it removed from 1.7 Release? @stnguyen90

**Comment by shlokstoira** [comment]
@LorenzoBloedow is there a way to self host this on my local machine, using docker and my own redis instead if Cloudflare workers and upstash redis. How can I go about doing that?

**Comment by xMrAfonso** [comment]
Indeed why was this removed? @stnguyen90

**Comment by muscaiu** [comment]
So is it possible to configure rate limits for a self-hosted Appwrite??

**Comment by Amitrochates** [comment]
@LorenzoBloedow Is borrow down currently?

**Comment by eldadfux** [comment]
UPDATE: The way we want to solve this is by introducing a global web application firewall, aka WAF. The idea is to allow each project, site or function to define its own custom rules.

Each rule will be defined by a set of condition which will result in action taken. Some of this actions will include:
- Bypass
- Deny
- Challenge
- RateLimit
- Redirect
- Log

Each action will allow you to set its own configuration like the target location for a redirect, or the timeframe, and count limits for the action `RateLimit`. Appwrite will be able to create conditions around your HTTP request (path, method, header etc) and will most likely also be able to introduce tighter integration with the platform with rules around (specific user, key, etc).

I have drafted a basic implementation of how the under the hood engine could look like. You can share feedback below.
https://github.com/utopia-php/waf
